mock/src/test_ctx.rs

```diff
@@ -0,0 +1,218 @@
+///         accs[2]
+///             .address(address!("0x000000000000000000000000000000000cafe002"))
+///             .balance(Word::from(1u64 << 30));
+///     },
+///     |mut txs, accs| {
+///         txs[0].to(accs[0].address).from(accs[2].address);
+///         txs[1]
+///             .to(accs[1].address)
+///             .from(accs[2].address)
+///             .nonce(Word::one());
+///     },
+///     |block, _tx| block.number(0xcafeu64),
+/// )
+/// .unwrap()
+/// .into();
+///
+/// // Now we can start generating the traces and items we need to inspect
+/// // the behaviour of the generated env.
+/// ```
+#[derive(Debug)]
+pub struct TestContext<const NACC: usize, const NTX: usize> {
+    /// chain id
+    pub chain_id: Word,
+    /// Account list
+    pub accounts: [Account; NACC],
+    /// history hashes contains most recent 256 block hashes in history, where
+    /// the lastest one is at history_hashes[history_hashes.len() - 1].
+    pub history_hashes: Vec<Word>,
+    /// Block from geth
+    pub eth_block: eth_types::Block<eth_types::Transaction>,
+    /// Execution Trace from geth
+    pub geth_traces: [eth_types::GethExecTrace; NTX],
+}
+
+impl<const NACC: usize, const NTX: usize> From<TestContext<NACC, NTX>> for GethData {
+    fn from(ctx: TestContext<NACC, NTX>) -> GethData {
+        GethData {
+            chain_id: ctx.chain_id,
+            history_hashes: ctx.history_hashes,
+            eth_block: ctx.eth_block,
+            geth_traces: ctx.geth_traces.to_vec(),
+            accounts: ctx.accounts.into(),
+        }
+    }
+}
+
+impl<const NACC: usize, const NTX: usize> TestContext<NACC, NTX> {
+    pub fn new<FAcc, FTx, Fb>(
+        history_hashes: Option<Vec<Word>>,
+        acc_fns: FAcc,
+        func_tx: FTx,
+        func_block: Fb,
+    ) -> Result<Self, Error>
+    where
+        FTx: FnOnce(Vec<&mut MockTransaction>, [MockAccount; NACC]),
+        Fb: FnOnce(&mut MockBlock, Vec<MockTransaction>) -> &mut MockBlock,
+        FAcc: FnOnce([&mut MockAccount; NACC]),
+    {
+        let mut accounts: Vec<MockAccount> = vec![MockAccount::default(); NACC];
+        let account_refs = accounts
+            .iter_mut()
+            .collect_vec()
+            .try_into()
+            .expect("Mismatched len err");
+        acc_fns(account_refs);
+        let accounts: [MockAccount; NACC] = accounts
+            .iter_mut()
+            .map(|acc| acc.build())
+            .collect_vec()
+            .try_into()
+            .expect("Mismatched acc len");
+
+        let mut transactions = vec![MockTransaction::default(); NTX];
+        let tx_refs = transactions.iter_mut().collect();
+        func_tx(tx_refs, accounts.clone());
+        let transactions: Vec<MockTransaction> =
+            transactions.iter_mut().map(|tx| tx.build()).collect();
+
+        let mut block = MockBlock::default();
+        block.transactions.extend_from_slice(&transactions);
+        func_block(&mut block, transactions).build();
```

Comment:
Why do we need to pass `transactions` if it's already in `block`? I saw it's never used so far.

Reply:
I see, I didn't notice that visibility issue. I am also good with current interface now, we can simplify it if it's indeed unnecessary in the future (removing is always easier).